### PR TITLE
Tv4g1 issue1414 local data source field

### PR DESCRIPTION
**Pull request by pdtouch**
### Field local source data
(Pull request on 5/4/2023 for tv4g1-issue1414-local-data-source-field)

## Issue #1414
# New Feature
## Description

Tripal Chado Analysis database table stores analytical methods to existing data.
The field Local Source Data helps in the Drupal/Tripal web interface to maintain
Analysis software content for Tripal 4.x, Drupal 9.5.

**Field files**

The PHP code to implement this field consists of 3 files in specific (PSR-4) directories
under the tripal/tripal_chado module as follows :

**Field type**
`tripal/tripal_chado/src/Plugin/Field/FieldType/ChadoLocalSourceDataDefault.php`
**Field widget**
`tripal/tripal_chado/src/Plugin/Field/Widget/ChadoLocalSourceDataWidgetDefault.php`
**Field Formatter**
`tripal/tripal_chado/src/Plugin/Field/Formatter/ChadoLocalSourceDataFormatterDefault.php`

## Testing?

Use this command to rebuild drush cache:

`export cntr_name=``sudo docker ps --format '{{.Names}}'`` ; docker exec -it $cntr_name drush cr `

Open Tripal web site on localhost : https://localhost:port

Screenshot 2023-04-26 at 2 12 57 PM

Login as user : drupaladmin / (password)

**View/Edit Fields for Analysis**

**Home -> Administration -> Structure -> Tripal Content Type ->
For Analysis content type under Label Column, click on
Edit -> Manage fields.**

Screenshot 2023-04-26 at 9 41 15 AM

It shows the fields currently active for Analysis content type

Screenshot 2023-04-26 at 9 42 00 AM

A Chado local analysis source data option should appear in Add Field screen.

Check Analysis content without local source data field

Enter these values in the form under :
**Content -> Tripal Content -> Add Tripal Content -> Analysis**

```
Name : Blast Analysis-1
Description :
Software : BLAST 2.9.0-1
Version Number : 2.9.0-1
Algorithm : BLASTX-1
URI : URI-1

```
Check the Postgres database table Analysis entry for the analysis_id = 1 (for record being tested)
Analysis DB Table Data after entering data. We see :

```
analysis_id | 1
name | Blast Analysis-1
description |
program | BLAST 2.9.0-1
programversion | 2.9.0-1
algorithm | BLASTX-1
sourcename |
sourceversion |
sourceuri | URI-1
timeexecuted | 2023-04-26 18:17:00.838846

```
Delete field URI  from existing in Analysis

**Home -> Administration -> Structure -> Tripal Content Type ->
For Analysis content type under Label Column, click on
Edit -> Manage fields. -> For URI field -> Edit -> Delete**

Screenshot 2023-04-26 at 1 38 38 PM

**Home -> Administration -> Structure -> Tripal Content Type ->
For Analysis content type under Label Column, click on
Edit -> Manage fields. -> For Version Number field -> Edit -> Delete**

Screenshot 2023-04-26 at 1 39 22 PM

**Add Chado local analysis source data field**

**Home -> Administration ->
Structure -> Tripal Content Type -> Edit Analysis -> Manage fields -> Select field type**

A 'Chado local analysis source data' option should appear in Add Field screen.

**Home -> Administration -> Structure -> Tripal Content Type ->
For Analysis content type under Label Column, click on
Edit -> Manage fields. -> Add**

Select the 'Chado local analysis source data' and give a Label as Data Source, Save and continue.

Screenshot 2023-04-26 at 1 44 15 PM

Save Field settings for Data Source

Screenshot 2023-04-26 at 1 45 36 PM

Check Analysis content 'with' local source data field

Enter these values in the form under :
**Content -> Tripal Content -> Add Tripal Content -> Analysis**

```
Name : Blast Analysis Test-2
Description :
Software : BLAST 2.9.0-2
Algorithm : BLASTX-2
Software version : 2.9.0-2
Local source URI : https://www.blast2.org/
Local Source Name :| local blast source name-2
Local Source Version : ls 2.9.0-2

```
Check the Postgres database table Analysis entry for the analysis_id = 2 (for record being tested)
Analysis DB Table Data after entering data. We see :

```
analysis_id | 2
name | Blast Analysis Test-2
description |
program | BLAST 2.9.0-2
programversion | 2.9.0-2
algorithm | BLASTX-2
sourcename | local blast source name-2
sourceversion | ls 2.9.0-2
sourceuri | https://www.blast2.org/

```
**sourcename, sourceuri and sourceversion column values are now populated in the Analysis table.**

### CODE REVIEW

**Comment by dsenalik** [comment]
We will move the colons from the Tripal 3 version of the formatter from the \<dd\> to the \<dt\>, so
20230505b

File: tripal_chado/src/Plugin/Field/FieldFormatter/ChadoLocalSourceDataFormatterDefault.php [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+ * @FieldFormatter(
+ *   id = "chado_local_source_data_formatter_default",
+ *   label = @Translation("Chado Local Source Data Formatter"),
+ *   description = @Translation("Analysis source name, version, and uri formatter"),
+ *   field_types = {
+ *     "chado_local_source_data_default"
+ *   }
+ * )
+ */
+class ChadoLocalSourceDataFormatterDefault extends ChadoFormatterBase {
```
Comment (spficklin) [inline tripal_chado/src/Plugin/Field/FieldFormatter/ChadoLocalSourceDataFormatterDefault.php]:
```suggestion
class ChadoSourceDataFormatterDefault extends ChadoFormatterBase {
```

File: tripal_chado/src/Plugin/Field/FieldType/ChadoLocalSourceDataDefault.php [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+ * @FieldType(
+ *   id = "chado_local_source_data_default",
+ *   label = @Translation("Chado Local Source Data"),
+ *   description = @Translation("The source and version of data used for this analysis"),
+ *   default_widget = "chado_local_source_data_widget_default",
+ *   default_formatter = "chado_local_source_data_formatter_default",
+ *   cardinality = 1,
+ * )
+ */
+class ChadoLocalSourceDataDefault extends ChadoFieldItemBase {
```
Comment (spficklin) [inline tripal_chado/src/Plugin/Field/FieldType/ChadoLocalSourceDataDefault.php]:
```suggestion
class ChadoSourceDataDefault extends ChadoFieldItemBase {
```

File: tripal_chado/src/Plugin/Field/FieldWidget/ChadoLocalSourceDataWidgetDefault.php [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+ * @FieldWidget(
+ *   id = "chado_local_source_data_widget_default",
+ *   label = @Translation("Chado Local Source Data Widget Default"),
+ *   description = @Translation("The default chado local source data widget which allows curators to manually enter analysis source on the content edit page."),
+ *   field_types = {
+ *     "chado_local_source_data_default"
+ *   }
+ * )
+ */
+class ChadoLocalSourceDataWidgetDefault extends ChadoWidgetBase {
```
Comment (spficklin) [inline tripal_chado/src/Plugin/Field/FieldWidget/ChadoLocalSourceDataWidgetDefault.php]:
```suggestion
class ChadoSourceDataWidgetDefault extends ChadoWidgetBase {
```

File: tripal_chado/src/Plugin/Field/FieldWidget/ChadoLocalSourceDataWidgetDefault.php [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+    $item_vals = $items[$delta]->getValue();
+
+    $elements = [];
+    $elements['record_id'] = [
+      '#type' => 'value',
+      '#default_value' => $item_vals['record_id'] ?? 0,
+    ];
+    $elements['sourceuri'] =  [
+      '#title' => "Local Source URI",
```
Comment (spficklin) [inline tripal_chado/src/Plugin/Field/FieldWidget/ChadoLocalSourceDataWidgetDefault.php]:
```suggestion
      '#title' => "Data Source URI",
```

File: tripal_chado/src/Plugin/Field/FieldWidget/ChadoLocalSourceDataWidgetDefault.php [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+      '#type' => 'textfield',
+      '#default_value' => $item_vals['sourceuri'] ?? '',
+    ];
+    $elements['sourcename'] =  [
+      '#title' => "Local Source Name",
```
Comment (spficklin) [inline tripal_chado/src/Plugin/Field/FieldWidget/ChadoLocalSourceDataWidgetDefault.php]:
```suggestion
      '#title' => "Data Source Name",
```

File: tripal_chado/src/Plugin/Field/FieldWidget/ChadoLocalSourceDataWidgetDefault.php [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+      '#type' => 'textfield',
+      '#default_value' => $item_vals['sourcename'] ?? '',
+    ];
+    $elements['sourceversion'] = [
+      '#type' => 'textfield',
+      '#default_value' => $item_vals['sourceversion'] ?? '',
+      '#title' => 'Local Source Version',
```
Comment (spficklin) [inline tripal_chado/src/Plugin/Field/FieldWidget/ChadoLocalSourceDataWidgetDefault.php]:
```suggestion
      '#title' => 'Data Source Version',
```

File: tripal_chado/src/Plugin/Field/FieldFormatter/ChadoLocalSourceDataFormatterDefault.php [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+use Drupal\Core\Field\FieldItemListInterface;
+use Drupal\Core\Link;
+use Drupal\Core\Url;
+use Drupal\tripal_chado\TripalField\ChadoFormatterBase;
+
+/**
+ * Plugin implementation of default Tripal string type formatter.
+ *
+ * @FieldFormatter(
+ *   id = "chado_local_source_data_formatter_default",
```
Comment (spficklin) [inline tripal_chado/src/Plugin/Field/FieldFormatter/ChadoLocalSourceDataFormatterDefault.php]:
```suggestion
 *   id = "chado_source_data_formatter_default",
```

File: tripal_chado/src/Plugin/Field/FieldFormatter/ChadoLocalSourceDataFormatterDefault.php [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+ *   label = @Translation("Chado Local Source Data Formatter"),
+ *   description = @Translation("Analysis source name, version, and uri formatter"),
+ *   field_types = {
+ *     "chado_local_source_data_default"
```
Comment (spficklin) [inline tripal_chado/src/Plugin/Field/FieldFormatter/ChadoLocalSourceDataFormatterDefault.php]:
```suggestion
 *     "chado_source_data_default"
```

File: tripal_chado/src/Plugin/Field/FieldFormatter/ChadoLocalSourceDataFormatterDefault.php [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+use Drupal\Core\Link;
+use Drupal\Core\Url;
+use Drupal\tripal_chado\TripalField\ChadoFormatterBase;
+
+/**
+ * Plugin implementation of default Tripal string type formatter.
+ *
+ * @FieldFormatter(
+ *   id = "chado_local_source_data_formatter_default",
+ *   label = @Translation("Chado Local Source Data Formatter"),
```
Comment (spficklin) [inline tripal_chado/src/Plugin/Field/FieldFormatter/ChadoLocalSourceDataFormatterDefault.php]:
```suggestion
 *   label = @Translation("Chado Data Source Formatter"),
```

File: tripal_chado/src/Plugin/Field/FieldType/ChadoLocalSourceDataDefault.php [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\tripal_chado\TripalField\ChadoFieldItemBase;
+use Drupal\tripal_chado\TripalStorage\ChadoIntStoragePropertyType;
+use Drupal\tripal_chado\TripalStorage\ChadoVarCharStoragePropertyType;
+
+/**
+ * Plugin implementation of Default Tripal field for sequence data.
+ *
+ * @FieldType(
+ *   id = "chado_local_source_data_default",
```
Comment (spficklin) [inline tripal_chado/src/Plugin/Field/FieldType/ChadoLocalSourceDataDefault.php]:
```suggestion
 *   id = "chado_source_data_default",
```

File: tripal_chado/src/Plugin/Field/FieldType/ChadoLocalSourceDataDefault.php [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+ *   label = @Translation("Chado Local Source Data"),
```
Comment (spficklin) [inline tripal_chado/src/Plugin/Field/FieldType/ChadoLocalSourceDataDefault.php]:
```suggestion
 *   label = @Translation("Chado Data Source"),
```

File: tripal_chado/src/Plugin/Field/FieldType/ChadoLocalSourceDataDefault.php [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+ *   description = @Translation("The source and version of data used for this analysis"),
+ *   default_widget = "chado_local_source_data_widget_default",
```
Comment (spficklin) [inline tripal_chado/src/Plugin/Field/FieldType/ChadoLocalSourceDataDefault.php]:
```suggestion
 *   default_widget = "chado_source_data_widget_default",
```

File: tripal_chado/src/Plugin/Field/FieldType/ChadoLocalSourceDataDefault.php [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+ *   default_formatter = "chado_local_source_data_formatter_default",
```
Comment (spficklin) [inline tripal_chado/src/Plugin/Field/FieldType/ChadoLocalSourceDataDefault.php]:
```suggestion
 *   default_formatter = "chado_source_data_formatter_default",
```

File: tripal_chado/src/Plugin/Field/FieldType/ChadoLocalSourceDataDefault.php [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+ *   cardinality = 1,
+ * )
+ */
+class ChadoLocalSourceDataDefault extends ChadoFieldItemBase {
+
+  public static $id = 'chado_local_source_data_default';
```
Comment (spficklin) [inline tripal_chado/src/Plugin/Field/FieldType/ChadoLocalSourceDataDefault.php]:
```suggestion
  public static $id = 'chado_source_data_default';
```

File: tripal_chado/src/Plugin/Field/FieldWidget/ChadoLocalSourceDataWidgetDefault.php [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+
+use Drupal\Core\Field\FieldItemListInterface;
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\tripal_chado\TripalField\ChadoWidgetBase;
+
+/**
+ * Plugin implementation of default Chado Sequence widget.
+ *
+ * @FieldWidget(
+ *   id = "chado_local_source_data_widget_default",
```
Comment (spficklin) [inline tripal_chado/src/Plugin/Field/FieldWidget/ChadoLocalSourceDataWidgetDefault.php]:
```suggestion
 *   id = "chado_source_data_widget_default",
```

File: tripal_chado/src/Plugin/Field/FieldWidget/ChadoLocalSourceDataWidgetDefault.php [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+ *   label = @Translation("Chado Local Source Data Widget Default"),
```
Comment (spficklin) [inline tripal_chado/src/Plugin/Field/FieldWidget/ChadoLocalSourceDataWidgetDefault.php]:
```suggestion
 *   label = @Translation("Chado Data Source Widget Default"),
```

File: tripal_chado/src/Plugin/Field/FieldWidget/ChadoLocalSourceDataWidgetDefault.php [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+ *   description = @Translation("The default chado local source data widget which allows curators to manually enter analysis source on the content edit page."),
```
Comment (spficklin) [inline tripal_chado/src/Plugin/Field/FieldWidget/ChadoLocalSourceDataWidgetDefault.php]:
```suggestion
 *   description = @Translation("The default source data widget which allows curators to manually enter analysis source data information on the content edit page."),
```

File: tripal_chado/src/Plugin/Field/FieldWidget/ChadoLocalSourceDataWidgetDefault.php [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+ *   field_types = {
+ *     "chado_local_source_data_default"
```
Comment (spficklin) [inline tripal_chado/src/Plugin/Field/FieldWidget/ChadoLocalSourceDataWidgetDefault.php]:
```suggestion
 *     "chado_source_data_default"
```

File: tripal_chado/src/Plugin/Field/FieldFormatter/ChadoLocalSourceDataFormatterDefault.php [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+ * )
+ */
+class ChadoLocalSourceDataFormatterDefault extends ChadoFormatterBase {
+
+  /**
+  * {@inheritdoc}
+  */
+  public function viewElements(FieldItemListInterface $items, $langcode) {
+    $elements = [];
+    $elements['#attached']['library'][] = 'tripal_chado/tripal_chado.field.ChadoLocalSourceDataFormatter';
```
Comment (spficklin) [inline tripal_chado/src/Plugin/Field/FieldFormatter/ChadoLocalSourceDataFormatterDefault.php]: I'm not sure what this library is?  Is it necessary?

**Review by spficklin — DISMISSED** [review]
I've made some suggested changes to take the word "Local" out of the class names, the field name and titles of widgets.  In Tripal v3 we put the vocabulary name at the front of all our field classes but in Tripal v4 we no longer need to do this.   You will need to rename the actual file names to match the class names (take out the word Local in the file name too).

**Comment by spficklin** [comment]
Thanks for submitting the field @pdtouch !  I just did a functional test and it worked just fine.  I also made some suggested changes to the code. Basically, the changes remove the word "local" from the class names, field names, and the widget titles.  In Tripal v3 we used to put the vocabulary name as a prefix to all our field classes, but this is no longer needed in Tripal v4.  

Can you make one other changes as well?  In Tripal v3 the 3 form elements were in a fieldset like this (note the box around all the fields with a title of "Data Source":

image

Can you reformat these fields to be layed out similarly?

**Comment by pdtouch** [comment]
All the requested changes are addressed, tested and found to work as expected. Thanks for the suggestions and help as how to make the changes.

**Comment by laceysanderson** [comment]
Thanks for the updates @pdtouch however, things seem to have gone missing in the process 🙈 I'm going to re-mark this as "waiting on submitter" until the field type and formatter files are recovered. That said, feel free to reach out to me on Slack and I can help with that process!

**Comment by dsenalik** [comment]
Testing today's version, I see this when I go to an analysis page
`
Warning: Undefined variable $delta in Drupal\tripal_chado\Plugin\Field\FieldFormatter\ChadoSourceDataFormatterDefault->viewElements() (line 51 of modules/contrib/tripal/tripal_chado/src/Plugin/Field/FieldFormatter/ChadoSourceDataFormatterDefault.php). `

The widget, however, is displaying perfectly as requested by @spficklin 
20230524_widget

**Comment by pdtouch** [comment]
Changes made as requested. 5/24/2023

**Comment by dsenalik** [comment]
I think that because of the cardinality of 1 we need to change how the formatter works. For an entity with no data source items looks like 
20230524_empty_items
and similar if there is content. So ~it is not an array~ (actually it is an iterable object), ~we can drop the loop and~ hard code delta as zero - let me experiment

**Comment by dsenalik** [comment]
I propose this, it looks like it is working for me
```
  public function viewElements(FieldItemListInterface $items, $langcode) {
    $elements = [];
    $content = '';

    foreach ($items as $delta => $item) {
      $sourcename_val = $item->get('sourcename')->getString() ;
      if ( !empty( $sourcename_val ) ) {
        $content .= "<dt>Source Name:</dt><dd>" . $sourcename_val . "</dd>";
      }
      $sourceversion_val = $item->get('sourceversion')->getString() ;
      if ( !empty( $sourceversion_val ) ) {
        $content .= "<dt>Source Version:</dt><dd>" . $sourceversion_val . "</dd>";
      }
      $sourceuri_val = $item->get('sourceuri')->getString() ;
      if ( !empty( $sourceuri_val ) ) {
        $url = $sourceuri_val;
        if (preg_match('|://|', $sourceuri_val)) {
          $url = Link::fromTextAndUrl($sourceuri_val, Url::fromUri($sourceuri_val, []))->toString();
        }
        $content .= "<dt>Source URI:</dt><dd>" . $url . "</dd>";
      }
    }

    if ($content) {
      $content = "<dl class=\"tripal-dl\">" . $content . "</dl>";
    }
    else {
      $content = 'The data source is not provided.';
    }

    // The cardinality of this field is always 1, so only create element for $delta of zero.
    $elements[0] = [
      '#type' => 'markup',
      '#markup' => $content,
    ];

    return $elements;
  }
```

**Comment by pdtouch** [comment]
Changed $delta to 0

**Review by dsenalik — APPROVED** [review]
I have tested this in a docker and it is working beautifully. I think it is ready for merging, and afterwards we can then migrate the three old fields to this new field in a separate issue.

**Testing.** I added this field to an analysis without an existing data source. The widget presented the following form when doing this, this is the desired appearance
20230525_addingfield

I then viewed the analysis. The three separate old fields that will be replaced by this field, plus the new field appeared as follows. This is desired behavior when there is no data source, based on the Tripal 3 field.
20230525_with-no-source

I then used the widget to add values, the widget has the enclosing box as @spficklin requested
20230525_widget-adding-values

And the final appearance
20230525_with-source

File: tripal_chado/src/Plugin/Field/FieldWidget/ChadoLocalSourceDataWidgetDefault.php (hunk 4 above)
Comment (dsenalik) [reply tripal_chado/src/Plugin/Field/FieldWidget/ChadoLocalSourceDataWidgetDefault.php]: These were changed to "Name", "Version", "URI"

File: tripal_chado/src/Plugin/Field/FieldFormatter/ChadoLocalSourceDataFormatterDefault.php (hunk 19 above)
Comment (dsenalik) [reply tripal_chado/src/Plugin/Field/FieldFormatter/ChadoLocalSourceDataFormatterDefault.php]: Library has been removed

**Pull request merged**